Incompressible flow example makeover (#313)
lethe-cfd/lethe: doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst

    set verbosity                              = verbose
  end

Simulation control
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The last subsection, which is generally the one we put at the top of the parameter files, is the ``simulation control`` . In this example, it is only used to specify the name of the output files:

.. code-block:: text

  subsection simulation control
    set method      = steady 
    set output name = output_cavity
  end

Running the simulation
----------------------
Launching the simulation is as simple as specifying the executable name and the parameter file. Assuming that the ``gls_navier_stokes_2d`` executable is within your path, the simulation can be launched by typing:

.. code-block:: text

  gls_navier_stokes_2d cavity.prm

Lethe will generate a number of files. The most important one bears the extension ``.pvd``. It can be read by popular visualization programs such as `Paraview <https://www.paraview.org/>`_. 

Base case results (Re=400)
---------------------------

Using Paraview, the steady-state velocity profile and the streamlines can be visualized:

.. image:: images/result.png
    :alt: velocity distribution
    :align: center

> It is also very interesting to compare the results with those obtained in the literature. A python script provided in the example folder allows to compare the velocity profile along de y axis for :math:`x=0.5` with results from the literature. Using this script, the following resuts are obtained for ``initial refinement = 6``

Reviewer: I would replace "de" by "the"

Higher-Reynolds case results (Re=7500)
---------------------------------------

We now consider the case at a Reynolds number of 7500. At this value of the Reynolds number, the ``steady`` solver will generally not converge as the problem is too non-linear (or too stiff). A workaround for this issue is to use an adjoint time-stepping strategy. This strategy consists in transforming the steady-state problem into a transient problem and to use an increasingly large time-step to reach a steady-state solution. This method is called ``steady_bdf`` in Lethe.

It can be used by modifying the ``simulation control`` subsection:

.. code-block:: text

  subsection simulation control
    set method                  = steady_bdf

    set adapt = true
    # Maximum CFL value
    set max cfl                      = 1000
    # Tolerance at which the simulation is stopped
    set stop tolerance               = 1e-10
  
    # Adaptative time step scaling
    set adaptative time step scaling = 1.05

    set time step = 0.001
  end

> The ``adapt`` parameter allows dynamic time-step adaptation. This feature is also used in transient simulations to carry simulations at a constant CFL number. The ``max cfl`` controls the maximum value of the CFL reached during the simulation. Remember that Lethe is an implicit solver and, as such, can theoretically manage absurdly large values of the CFL. The ``stop-tolerance`` controls the initial tolerance of a time-step below which steady-state will be considered reached.  The ``time step`` controls the initial value of the time step and, finally, the ``adaptative time step scaling`` controls the rate of increase of the time step. The increase of the time step follows:

Reviewer: In the 4th sentence, "stop-tolerance" should be replaced by "stop tolerance"

==================================
2D Lid-Driven Cavity Flow
==================================

This is the first Lethe example. It showcases a classical fluid mechanics problem, the lid-driven cavity. This example also introduces the concept of parameter files to parametrize Lethe simulations.

Features used in this example
----------------------------------
- Solver : ``gls_navier_stokes_2d``, ``gd_navier_stokes_2d`` (with Q2-Q1)
- Steady-state problem
- Setting initial conditions
- Adjoint time-stepping for steady-state problems

Location of the examples
------------------------

> - Base case (Re=400) : ``/examples/incompressible_flows/cavity.prm``

Reviewer 2: Maybe use inline math for Re?
```suggestion
- Base case (:math:`R_e=400`) : ``/examples/incompressible_flows/cavity.prm``
```

> - Re=7500 : ``/examples/incompressible_flows/Reynolds_7500/cavity.prm``

Reviewer 2: Path could be a link to the mentioned file on GitHub

Author: Good question. I'm not sure which is better here. I guess specifying the path will enable users to find quickly the example they are looking for and run it. Pointing to the github repo would have a similar impact, but would be a bit more indirect I guess.

Description of the case
-----------------------

> The lid-driven cavity is a classical fluid dynamics problem . It consists in the flow in an enclosed cavity, where one of the wall (in this example, the top wall) is put in motion at a constant tangential velocity. This case is often used to benchmark the capacity of CFD software because of the simplicity of it's geometry and the availabily of multiple reference results over a large range of Reynolds number. In this example, we investigate this problem in 2D, but simulating this case in 3D would require only a few changes. The geometry and boundary conditions are illustrated in the following figure:

Reviewer 3: ```suggestion
The lid-driven cavity is a classical fluid dynamics problem . It consists in the flow in an enclosed cavity, where one of the walls (in this example, the top wall) is put in motion at a constant tangential velocity. This case is often used to benchmark the capacity of CFD software because of the simplicity of its geometry and the availability of multiple reference results over a large range of Reynolds numbers. In this example, we investigate this problem in 2D, but simulating this case in 3D would require only a few changes. The geometry and boundary conditions are illustrated in the following figure:
```

.. image:: images/geo.png
    :alt: The geometry and boundary conditions
    :align: center
    :name: geometry

> Only the upper wall boundary moves in the x direction with a constant velocity (u = 1 m/s) while the other boundaries are static. We will discuss later in this example the meaning of the ID of the boundary conditions used in this figure.

Reviewer 2: Inline math for u = 1 m/s?

> We first investigate this case at a Reynolds number of 400 for which a steady-state solution can be easily obtained. Later, we will increase this Reynolds number to 7500 and leverage some of the advanced functionnalities of Lethe to reach a steady-state solution. 

Reviewer 3: ```suggestion
We first investigate this case at a Reynolds number of 400 for which a steady-state solution can be easily obtained. Later, we will increase this Reynolds number to 7500 and leverage some of the advanced functionalities of Lethe to reach a steady-state solution. 
```

Parameter file
--------------

Lethe simulations are controlled by *parameter files* which possess the extension ``.prm``. This is the default text format of the ParameterHandler class of the deal.ii library from which Lethe derives. For more information on this class, we refer to the `deal.II documentation <https://www.dealii.org/current/doxygen/deal.II/classParameterHandler.html>`_. 

> Parameter files are made of subsections which describe a portion of the simulation (e.g. mesh generation, physical properties, simulation control). In parameter files, line starting with ``#`` are comments. Parameters are set using the following syntax:

Reviewer 3: ```suggestion
Parameter files are made of subsections which describe a portion of the simulation (e.g. mesh generation, physical properties, simulation control). In parameter files, lines starting with ``#`` are comments. Parameters are set using the following syntax:
```

.. code-block:: text

    set parameter name = value

The syntax is flexible. Parameters do not need to be specified in a specific order, but only within the subsection in which they belong. For a full list of the parameters within Lethe, we refer to the parameter page.

To set-up the lid-driven cavity case, we first need to establish the mesh used for the simulation.

Mesh
~~~~~

The ``mesh`` subsection specifies the computational grid:

.. code-block:: text

    subsection mesh
        set type                 = dealii
        set grid type            = hyper_cube
        set grid arguments       = 0 : 1 : true
        set initial refinement   = 6
    end

> The ``type`` specifies the mesh format used. At the moment, Lethe supports two mesh format: ``dealii`` and ``gmsh``. ``dealii`` meshes are in-situ generated meshes for simple geometries. The type of grid generated is specified by the ``grid type`` parameters and this grid is parametrized by it's ``grid arguments``. We refer to the documentation of the deal.ii `GridGenerator <https://www.dealii.org/current/doxygen/deal.II/namespaceGridGenerator.html>`_ for a detailed explanation of the available grids. 

Reviewer 3: ```suggestion
The ``type`` specifies the mesh format used. At the moment, Lethe supports two mesh formats: ``dealii`` and ``gmsh``. ``dealii`` meshes are in-situ generated meshes for simple geometries. The type of grid generated is specified by the ``grid type`` parameters and this grid is parametrized by it's ``grid arguments``. We refer to the documentation of the deal.ii `GridGenerator <https://www.dealii.org/current/doxygen/deal.II/namespaceGridGenerator.html>`_ for a detailed explanation of the available grids. 
```

> Since the lid-driven cavity problem domain is a square, we use the *hyper_cube* ``grid_type``. The arguments of this grid type are the position of the bottom left corner, the position of the top right corner and the option to colorize the boundaries in order to give each of them a unique ID. The IDs will be used to set the boundary conditions on specific part of the boundary of the domain. The ID given to each face were given in the graphical description of the case. If ``colorize`` option were set to false, all boundaries would have been given the ID ``0``.

Reviewer 3: ```suggestion
Since the lid-driven cavity problem domain is a square, we use the *hyper_cube* ``grid_type``. The arguments of this grid type are the position of the bottom left corner, the position of the top right corner and the option to colorize the boundaries in order to give each of them a unique ID. The IDs will be used to set the boundary conditions on specific parts of the boundary of the domain. The ID given to each face was given in the graphical description of the case. If ``colorize`` option were set to false, all boundaries would have been given the ID ``0``.
```

> It is a bit surprising that the position of the bottom left and the top right corner are specified by a single value. Since the geometry is a square, the positions of the corner is specified using a single number, assuming that this identifies both the x and y value associated with that point. Other grid generators, such as the ``hyper_rectangle``, allow for more flexibility.

Reviewer 3: ```suggestion
It is a bit surprising that the position of the bottom left and the top right corner are specified by a single value. Since the geometry is a square, the position of the corner is specified using a single number, assuming that this identifies both the x and y value associated with that point. Other grid generators, such as the ``hyper_rectangle``, allow for more flexibility.
```

> The last parameter specifies the ``initial refinement`` of the grid. Most deal.ii grid generators contain a minimal number of cells. For example, the *hyper_cube* mesh is made of a single cell. Indicating an `initial refinement=6` implies that the initial mesh is refined 6 times. In 2D, each cell is divided by 4 per refinement. Consequently, the final grid is made of :math:`2^{(2\cdot6)}=4096` cells.

Reviewer 2: Double backtick for monospace

```suggestion
The last parameter specifies the ``initial refinement`` of the grid. Most deal.ii grid generators contain a minimal number of cells. For example, the *hyper_cube* mesh is made of a single cell. Indicating an ``initial refinement=6`` implies that the initial mesh is refined 6 times. In 2D, each cell is divided by 4 per refinement. Consequently, the final grid is made of :math:`2^{(2\cdot6)}=4096` cells.
```

Boundary conditions
~~~~~~~~~~~~~~~~~~~

> The ``boundary conditions`` subsection establishes the boundary condition:

Reviewer 3: ```suggestion
The ``boundary conditions`` subsection establishes the constraints on different parts of the domain:
```

.. code-block:: text

    subsection boundary conditions
    set number                  = 4
        subsection bc 0
            set id                = 0
            set type              = noslip
        end
        subsection bc 1
            set id                = 1
            set type              = noslip
        end
        subsection bc 2
            set id                = 2
            set type              = noslip
        end
        subsection bc 3
            set id                = 3
            set type              = function
            subsection u
                set Function expression = 1
            end
            subsection v
                set Function expression = 0
            end
        end
    end

> First, the ``number`` of boundary conditions to be applied must be specified. For each boundary condition, the ``id`` of the boundary as well as its ``type`` must be specified. The left (``0``), right (``1``) and bottom (``2``) walls are static and, consequently, a ``noslip`` boundary condition can be used. This boundary condition imposes :math:`\mathbf{u} = [0,0]^T`. For the top wall, we use the `function` boundary type. This type of boundary condition allows us to define the value of the velocity components using `Function expression`. We set :math:`u=1` and :math:`v=0`. Note that the `Function expression` supports writing complex mathematical expression which may depend on the spatial coordinates (:math:`x,y,z`) and on time.

Reviewer 2: Double backtick for monospace
```suggestion
First, the ``number`` of boundary conditions to be applied must be specified. For each boundary condition, the ``id`` of the boundary as well as its ``type`` must be specified. The left (``0``), right (``1``) and bottom (``2``) walls are static and, consequently, a ``noslip`` boundary condition can be used. This boundary condition imposes :math:`\mathbf{u} = [0,0]^T`. For the top wall, we use the ``function`` boundary type. This type of boundary condition allows us to define the value of the velocity components using ``Function expression``. We set :math:`u=1` and :math:`v=0`. Note that the ``Function expression`` supports writing complex mathematical expression which may depend on the spatial coordinates (:math:`x,y,z`) and on time.
```

> First, the ``number`` of boundary conditions to be applied must be specified. For each boundary condition, the ``id`` of the boundary as well as its ``type`` must be specified. The left (``0``), right (``1``) and bottom (``2``) walls are static and, consequently, a ``noslip`` boundary condition can be used. This boundary condition imposes :math:`\mathbf{u} = [0,0]^T`. For the top wall, we use the `function` boundary type. This type of boundary condition allows us to define the value of the velocity components using `Function expression`. We set :math:`u=1` and :math:`v=0`. Note that the `Function expression` supports writing complex mathematical expression which may depend on the spatial coordinates (:math:`x,y,z`) and on time.

Reviewer 3: ```suggestion
First, the ``number`` of boundary conditions to be applied must be specified. For each boundary condition, the ``id`` of the boundary as well as its ``type`` must be specified. The left (``0``), right (``1``) and bottom (``2``) walls are static and, consequently, a ``noslip`` boundary condition can be used. This boundary condition imposes :math:`\mathbf{u} = [0,0]^T`. For the top wall, we use the `function` boundary type. This type of boundary condition allows us to define the value of the velocity components using `Function expression`. We set :math:`u=1` and :math:`v=0`. Note that the `Function expression` supports writing complex mathematical expressions which may depend on the spatial coordinates (:math:`x,y,z`) and on time.
```

Physical properties
~~~~~~~~~~~~~~~~~~~

For the base case, we wish to simulate the lid-driven cavity at a Reynolds number of 400. Since the characteristic dimension of the cavity is :math:`L=1` and the velocity of the top boundary is :math:`u=1`, the Reynolds number is :math:`Re=\frac{1}{\nu}` where :math:`\nu` is the kinematic viscosity. The kinematic viscosity is set by the ``physical properties`` subsection:

.. code-block:: text

  subsection physical properties
    subsection fluid 0
      set kinematic viscosity            = 0.0025
    end
  end

By default, simulations only contain a single fluid which is labeled ``0``.

FEM interpolation
~~~~~~~~~~~~~~~~~

Lethe supports the use of arbitrary interpolation order. The default solver for this case is ``gls_navier_stokes_2d`` which uses a stabilized method and supports equal order interpolation. 

We specify the interpolation order for both pressure and velocity using the ``FEM`` subsection:

.. code-block:: text

    subsection FEM
        set velocity order            = 1
        set pressure order            = 1
    end

.. warning:: 

>     An alternative would be to use the ``gd_navier_stokes_2d`` solver for `LBB <https://en.wikipedia.org/wiki/Ladyzhenskaya%E2%80%93Babu%C5%A1ka%E2%80%93Brezzi_condition>`_ stable elements must be used (e.g. Qn-Q(n-1)). Only the stabilized solver supports the use of equal order elements. 

Reviewer 3: ```suggestion
    An alternative would be to use the ``gd_navier_stokes_2d`` solver; for `LBB <https://en.wikipedia.org/wiki/Ladyzhenskaya%E2%80%93Babu%C5%A1ka%E2%80%93Brezzi_condition>`_ stable elements must be used (e.g. Qn-Q(n-1)). Only the stabilized solver supports the use of equal order elements. 
```

Non-linear solver parameters
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Lethe is an implicit CFD solver. Consequently, each time-step requires the solution of a non-linear system of equations. By default, Lethe uses a Newton solver for which a ``tolerance`` must be specified:

.. code-block:: text

  subsection non-linear solver
    set tolerance               = 1e-8
    set verbosity               = verbose
  end

> The ``verbosity`` option specifies if details about the non-linear solver steps (residual value and iteration number) will be printed out to the terminal. By setting it to ``verbose``, this information is printed out, wheras ``quiet`` would mute all outputs of the non-linear solver. We recommend to always set ``verbosity=verbose`` in order to monitor possible non-convergence of the solver.

Reviewer 3: ```suggestion
The ``verbosity`` option specifies if details about the non-linear solver steps (residual value and iteration number) will be printed out to the terminal. By setting it to ``verbose``, this information is printed out, whereas ``quiet`` would mute all outputs of the non-linear solver. We recommend to always set ``verbosity=verbose`` in order to monitor possible non-convergence of the solver.
```

Linear solver parameters
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Each non-linear solver step requires the solution of a linear system of equations. Lethe has multiple options to parametrize it's linear solvers, but here, we only enable it's verbosity to monitor the number of iteration per time step and use the default parameters for the rest. 

Reviewer 3: ```suggestion
Each non-linear solver step requires the solution of a linear system of equations. Lethe has multiple options to parametrize its linear solvers, but here, we only enable its verbosity to monitor the number of iteration per time step and use the default parameters for the rest. 
```

.. note:: 

>     A good thing to remember is that, generally, linear solver parameter control the robustness of the simulation.

Reviewer 3: ```suggestion
    A good thing to remember is that, generally, linear solver parameters control the robustness of the simulation.
```

.. code-block:: text

  subsection linear solver
    set method                                 = amg
    set verbosity                              = verbose
  end

Simulation control
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The last subsection, which is generally the one we put at the top of the parameter files, is the ``simulation control`` . In this example, it is only used to specify the name of the output files:

.. code-block:: text

  subsection simulation control
    set method      = steady 
    set output name = output_cavity
  end

Running the simulation
----------------------
Launching the simulation is as simple as specifying the executable name and the parameter file. Assuming that the ``gls_navier_stokes_2d`` executable is within your path, the simulation can be launched by typing:

.. code-block:: text

  gls_navier_stokes_2d cavity.prm

Lethe will generate a number of files. The most important one bears the extension ``.pvd``. It can be read by popular visualization programs such as `Paraview <https://www.paraview.org/>`_. 

> Results at Re=400

Reviewer 3: ```suggestion
Base case results (Re=400)
```

----------------------

Using Paraview, the steady-state velocity profile and the streamlines can be visualized:

.. image:: images/result.png
    :alt: velocity distribution
    :align: center

It is also very interesting to compare the results with those obtained in the literature. A python script provided in the example folder allows to compare the velocity profile along de y axis for :math:`x=0.5` with results from the literature. Using this script, the following resuts are obtained for ``initial refinement = 6``

.. image:: images/lethe_ghia_re_400_comparison.png
    :alt: re_400_comparison
    :align: center

> We note that the agreement is perfect. This is not surprising, especially considering that these results were obtained at relatively low Reynolds number.

Reviewer 3: ```suggestion
We note that the agreement is perfect. This is not surprising, especially considering that these results were obtained at a relatively low Reynolds number.
```

.. note:: 

>     The vtu files generated by Lethe are compressed archives. Consequently, they cannot be postprocessed directly. Although they can be easily post-processed using Paraview, it is sometimes necessary to be able to work with the raw data. The python library pyvista allows us to do this.

Reviewer 3: ```suggestion
    The vtu files generated by Lethe are compressed archives. Consequently, they cannot be postprocessed directly. Although they can be easily post-processed using Paraview, it is sometimes necessary to be able to work with the raw data. The python library `PyVista <https://www.pyvista.org/>`_  allows us to do this.
```

> Case at Re=7500

Reviewer 3: ```suggestion
Higher-Reynolds case results (Re=7500)
```

.. math::

  \Delta t_{n+1} = \alpha \Delta t_{n}

where :math:`\alpha` is the ``adaptative time step scaling`` .

A complimentary strategy is to also start the problem from a closer initial condition. This is achieved with the ``initial conditions`` subsection:

.. code-block:: text

  subsection initial conditions
    set type      = viscous
    # viscosity for viscous initial conditions
    set viscosity = 0.1
  end

Here we use ``viscous`` initial condition in which an auxiliary steady-state problem at a higher viscosity is used to generate a more realistic initial condition to the problem at hand.

> Results

Reviewer 3: I would delete this and keep only one subsection for the higher Reynolds case

Author: You are right, it detracts the focus a bit. I'll remove that part. We can leave it to another case.

.. image:: images/result_re_7500.png
    :alt: velocity distribution
    :align: center

We see that new recirculation regions occur in the flow.

Using a similar python script, we can compare the results obtained with those from the literature. We note that there is quite a good agreement with the results from the literature, even in the top region where the velocity field changes abruptly.

.. image:: images/re_7500_mesh_7/lethe_ghia_re_7500_comparison.png
    :alt: re_7500_comparison_mesh_7
    :align: center

.. image:: images/re_7500_mesh_7/lethe_ghia_re_7500_comparison_zoom.png
    :alt: re_7500_comparison_mesh_7
    :align: center

Increasing the number of cells by a factor 4 (to ~65k) allows for an even slightly better agreement.

.. image:: images/re_7500_mesh_8/lethe_ghia_re_7500_comparison.png
    :alt: re_7500_comparison_mesh_8
    :align: center

.. image:: images/re_7500_mesh_8/lethe_ghia_re_7500_comparison_zoom.png
    :alt: re_7500_comparison_mesh_8
    :align: center

Possibilities for extension
----------------------------

> - **Validate at higher Reynolds number:** The Erturk 2005 data within the example investigates this case up to a Reynolds number of 20000. Using the adjoint time-stepping ``steady_bdf`` scheme, it is an interesting exercise to simulate these more complex cases.

Reviewer 3: ```suggestion
- **Validate at even higher Reynolds numbers:** The Erturk 2005 data within the example investigates this case up to a Reynolds number of 20000.  It is an interesting exercise to simulate these more complex cases using the adjoint time-stepping ``steady_bdf`` scheme. 
```

> - **High-order methods:** Lethe supports higher order interpolation. This can yield to much better results with an equal number of degrees of freedom than traditional second-order (Q1-Q1) methods, especially at higher Reynolds number. 

Reviewer 3: ```suggestion
- **High-order methods:** Lethe supports higher order interpolation. This can yield much better results with an equal number of degrees of freedom than traditional second-order (Q1-Q1) methods, especially at higher Reynolds numbers. 
```

> - **Dynamic mesh adaptation:** Lethe supports dynamic mesh adaptation. Running this case with dynamic mesh adaptation could potentially yield better results.s

Reviewer 3: ```suggestion
- **Dynamic mesh adaptation:** Lethe supports dynamic mesh adaptation. Running this case with dynamic mesh adaptation could potentially yield better results.
```
